File: backend/app/api/platforms.py
# 平台管理 API
import os
import shutil
from pathlib import Path
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime

from app.store import store
from app.core.config import settings
from app.platforms.base import PlatformType, ContentType, PublishStatus
from app.platforms import supported_platforms

router = APIRouter()


VALID_PLATFORMS = {p.value for p in PlatformType}


# ============ Pydantic 模型 ============

class AccountCreate(BaseModel):
    platform: str
    name: str = Field(..., min_length=1, max_length=100)
    account_id: Optional[str] = None  # 平台上的账号 ID
    description: str = ""
    cookie_path: Optional[str] = None  # Phase 2: sau 的 storage_state 路径


class AccountUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    status: Optional[str] = None  # active/inactive
    description: Optional[str] = None
    cookie_path: Optional[str] = None


class PublishRequest(BaseModel):
    content_id: str
    platform: str
    scheduled_time: Optional[datetime] = None
    account_id: Optional[str] = None  # Phase 2: 多账号
    video_id: Optional[str] = None    # Phase 2: 关联视频


class PublishNowRequest(BaseModel):
    """前端 "立即发布视频" 按钮"""
    content_id: str
    platform: str
    account_id: str
    video_id: str


# ============ 路由 ============

@router.get("/accounts")
async def list_accounts():
    """列出所有平台账号"""
    return store.list_accounts()


@router.post("/accounts", status_code=201)
async def add_account(req: AccountCreate):
    """添加平台账号"""
    if req.platform not in VALID_PLATFORMS:
        raise HTTPException(status_code=400, detail=f"Unsupported platform: {req.platform}")
    data = req.model_dump()
    # cookie_path 缺省 = storage/cookies/{name}.json
    if not data.get("cookie_path"):
        data["cookie_path"] = str(Path(settings.COOKIES_DIR) / f"{data['name']}.json")
    return store.add_account(data)


@router.put("/accounts/{account_id}")
async def update_account(account_id: str, update: AccountUpdate):
    """更新平台账号(用 store.update_account,行为一致)"""
    payload = update.model_dump(exclude_unset=True)
    acc = store.update_account(account_id, payload)
    if not acc:
        raise HTTPException(status_code=404, detail="Account not found")
    return acc


@router.delete("/accounts/{account_id}")
async def delete_account(account_id: str):
    """删除平台账号"""
    if not store.delete_account(account_id):
        raise HTTPException(status_code=404, detail="Account not found")
    return {"message": "deleted", "id": account_id}


@router.post("/publish", status_code=201)
async def publish_content(req: PublishRequest):
    """发布内容到平台（创建发布记录;Phase 2 可选绑 account + video）"""
    if req.platform not in VALID_PLATFORMS:
        raise HTTPException(status_code=400, detail=f"Unsupported platform: {req.platform}")

    content = store.get_content(req.content_id)
    if not content:
        raise HTTPException(status_code=404, detail="Content not found")

    # 优先用请求里的 video_id, 否则用 content.video_id
    video_id = req.video_id or content.get("video_id")

    record = {
        "content_id": req.content_id,
        "content_title": content.get("title"),
        "platform": req.platform,
        "scheduled_time": req.scheduled_time.isoformat() if req.scheduled_time else None,
        "status": "scheduled" if req.scheduled_time else "pending",
        "account_id": req.account_id,
        "video_id": video_id,
    }
    record = store.add_publish_record(record)

    # 定时发布:同时建调度任务
    if req.scheduled_time:
        store.add_scheduled_task({
            "content_id": req.content_id,
            "platform": req.platform,
            "scheduled_time": req.scheduled_time.isoformat(),
            "status": "pending",
            "publish_id": record["publish_id"],
            "account_id": req.account_id,
            "video_id": video_id,
        })

    return record


@router.post("/publish-now")
async def publish_now(req: PublishNowRequest):
    """前端"立即发布"按钮 — 同步派发,返回 dispatch summary。
    不走 scheduler tick,直接调 scheduler_loop.publish_now。
    """
    if req.platform not in VALID_PLATFORMS:
        raise HTTPException(status_code=400, detail=f"Unsupported platform: {req.platform}")
    if PlatformType(req.platform) not in supported_platforms():
        raise HTTPException(
            status_code=501,
            detail=f"platform {req.platform} 暂未实装 adapter。已支持: "
                   f"{[p.value for p in supported_platforms()]}",
        )
    from app.services.scheduler_loop import scheduler_loop
    result = await scheduler_loop.publish_now(
        content_id=req.content_id,
        platform=req.platform,
        account_id=req.account_id,
        video_id=req.video_id,
    )
    return result


@router.get("/publish/records")
async def list_publish_records(status: Optional[str] = None):
    """列出发布记录"""
    return store.list_publish_records(status=status)


@router.get("/status/{publish_id}")
async def get_publish_status(publish_id: str):
    """获取发布状态"""
    record = store.get_publish_record(publish_id)
    if not record:
        raise HTTPException(status_code=404, detail="Publish record not found")
    return record


@router.get("/sau-status")
async def sau_status(account_id: Optional[str] = None):
    """健康检查: sau 是否装、版本、默认 cookie 路径等。
    帮前端 Settings / PublishRecords 提示用户该跑什么命令。
    """
    from app.platforms.douyin import _resolve_sau_bin
    sau_bin = _resolve_sau_bin(getattr(settings, "DOUYIN_SAU_BIN", "sau"))
    info: Dict[str, Any] = {
        "sau_installed": bool(sau_bin),
        "sau_bin": sau_bin,
        "videos_dir": settings.VIDEOS_DIR,
        "cookies_dir": settings.COOKIES_DIR,
        "default_account": settings.DOUYIN_DEFAULT_ACCOUNT,
        "supported_platforms": [p.value for p in supported_platforms()],
        "login_command": (
            f"{sau_bin} douyin login --account {settings.DOUYIN_DEFAULT_ACCOUNT}"
            if sau_bin else None
        ),
    }
    if account_id:
        acc = store.get_account(account_id)
        if acc:
            cookie_path = acc.get("cookie_path") or ""
            info["account_id"] = acc.get("id")
            info["account_name"] = acc.get("name")
            info["account_platform"] = acc.get("platform")
            info["cookie_path"] = cookie_path
            info["cookie_exists"] = bool(cookie_path) and Path(cookie_path).is_file()
    else:
        # 默认账号的 cookie 状态
        default_cookie = str(Path(settings.COOKIES_DIR) / f"{settings.DOUYIN_DEFAULT_ACCOUNT}.json")
        info["cookie_path"] = default_cookie
        info["cookie_exists"] = Path(default_cookie).is_file()
    return info

File: frontend/src/api/platforms.ts
// 平台 API 客户端
import request from './index'

// ---- 共享类型 ----

export interface PlatformAccount {
  id: string
  platform: string
  name: string
  account_id?: string
  description?: string
  cookie_path?: string       // Phase 2: sau 的 storage_state 路径
  status: 'active' | 'inactive'
  created_at: string
}

export type PublishStatus =
  | 'pending'
  | 'scheduled'
  | 'uploading'
  | 'publishing'           // P0-1: 公众号派发中
  | 'draft_added'          // P0-1: 公众号草稿已写入(API 流里内部态)
  | 'freepublish_submitted'// P0-1: 公众号派发已提交,等待审核
  | 'published'
  | 'failed'

export interface PublishRecord {
  publish_id: string
  content_id: string
  content_title?: string
  platform: string
  status: PublishStatus | string  // 服务端可能扩展新值
  scheduled_time?: string | null
  url?: string | null              // Phase 2: 平台侧发布 URL
  platform_publish_id?: string | null
  error_message?: string | null
  attempted_at?: string | null
  video_id?: string | null
  account_id?: string | null
  // P0-1 公众号新字段
  article_url?: string | null       // 公众号文章 URL(同 url,但语义明确)
  draft_media_id?: string | null    // 草稿 id
  freepublish_id?: string | null    // 派发 id
  freepublish_status?: number | null // 0=成功 1=审核中 2=原创 3=参数 4=失败
  thumb_media_id?: string | null    // 永久素材 id(封面)
  created_at: string
  executed_at?: string
}

export interface PublishNowParams {
  content_id: string
  platform: string
  account_id: string
  video_id: string
}

export interface PublishNowResponse {
  status: 'published' | 'failed' | 'retrying'
  task_id?: string
  publish_id?: string
  platform_publish_id?: string
  url?: string
  error?: string
}

// P0-1: 公众号全自动图文发布参数/响应
export interface PublishWechatParams {
  content_id: string
  account_id: string
}

export interface PublishWechatResponse {
  status: 'published' | 'failed' | 'freepublish_submitted'
  url?: string | null                // 公众号文章 URL(成功时)
  platform_publish_id?: string | null
  draft_media_id?: string | null
  freepublish_id?: string | null
  freepublish_status?: number | null
  thumb_media_id?: string | null
  error_message?: string | null
  body_html?: string                 // 改写后的 HTML(调试用)
}

export interface SauStatus {
  sau_installed: boolean
  sau_bin: string | null
  videos_dir: string
  cookies_dir: string
  default_account: string
  supported_platforms: string[]
  login_command: string | null
  cookie_path?: string
  cookie_exists?: boolean
  account_id?: string
  account_name?: string
  account_platform?: string
}

// ---- API ----

export const platformApi = {
  // 账号
  listAccounts: () => request.get<PlatformAccount[], PlatformAccount[]>('/platforms/accounts'),
  addAccount: (data: { platform: string; name: string; account_id?: string; description?: string; cookie_path?: string }) =>
    request.post<PlatformAccount, PlatformAccount>('/platforms/accounts', data),
  updateAccount: (id: string, data: Partial<PlatformAccount>) =>
    request.put<PlatformAccount, PlatformAccount>(`/platforms/accounts/${id}`, data),
  deleteAccount: (id: string) =>
    request.delete<{ message: string; id: string }>(`/platforms/accounts/${id}`),

  // 发布
  publish: (data: { content_id: string; platform: string; scheduled_time?: string; account_id?: string; video_id?: string }) =>
    request.post<PublishRecord, PublishRecord>('/platforms/publish', data),
  publishVideo: (data: PublishNowParams) =>
    request.post<PublishNowResponse, PublishNowResponse>('/platforms/publish-now', data),
  // P0-1: 公众号全自动图文混排发布
  publishWechatArticle: (data: PublishWechatParams) =>
    request.post<PublishWechatResponse, PublishWechatResponse>('/platforms/publish-article-now', data),
  listPublishRecords: (status?: string) =>
    request.get<PublishRecord[], PublishRecord[]>('/platforms/publish/records', { params: { status } }),
  getPublishStatus: (id: string) =>
    request.get<PublishRecord, PublishRecord>(`/platforms/status/${id}`),

  // Phase 2 诊断
  sauStatus: (accountId?: string) =>
    request.get<SauStatus, SauStatus>('/platforms/sau-status', { params: accountId ? { account_id: accountId } : {} }),
}
